### ISSUE TITLE
Refactor `Env` (and `Agent`?) to be more functional and return a `state` like in JaxMARL and Jax+RL

### ISSUE THREAD

**Issue by conorheins** [infer-actively/pymdp#273]
Raised by @tverbele 

In the pymdp jax Env we currently keep the env state in the Env object and after rollout we get a batched version of this Env object back.
```python
observation_next, env_next = env.step(
            rng_key=keys[2:], actions=action_next
        ) 
```
Was looking a bit in how jax+RL is doing it, and for instance in gymnax / jaxMARL the flow is typically:

```python
key = jax.random.PRNGKey(0)
key, key_reset, key_act, key_step = jax.random.split(key, 4)

# Instantiate the environment & its settings.
env, env_params = gymnax.make("Pendulum-v1")

# Reset the environment.
obs, state = env.reset(key_reset, env_params)

# Sample a random action.
action = env.action_space(env_params).sample(key_act)

# Perform the step transition.
n_obs, n_state, reward, done, _ = env.step(key_step, state, action, env_params)
```

where the state is typically a separate object that gets passed (and that would end up in the info dict)...
Is there a rationale to keep everything packed in the Env equinox module, or should we consider moving to something more like these other jax environments use to be consistent?

**Comment by Arun-Niranjan** [comment]
I think this refactor makes sense to do. It seems like it would have the following benefits:
1. more consistent with other JAX environments (as noted above)
2. testing our environment functions will in theory be easier as they will have fewer side-effects

**Comment by tverbele** [comment]
From a pymdp / active inference perspective, I think it makes sense to have an Environment API like

```
obs, state = env.reset(key)
obs, state = env.step(key, state, action)
```
where `obs` would be an array of jnp arrays that you can feed into `Agent.infer_states` and `state` something environment specific. We could also provide wrapper implementations for some of the above environments to test active inference agents on these RL envs (e.g. converting the reward into an observation modality etc)

However, a few design decisions to make are:
- do we provide a key each call of `step` or let the env handle it (our current API does the former which lgtm)?
- do we consider a `done` signal and/or should the environment "auto-reset"?
- do we provide a way of specifying the goal in the environment, i.e. by specifying a C tensor or some reward signal after all?
- do we return `obs` or rather a more expressive `Timestep` dataclass object?
- do we provide a standard `render` method , i.e. `env.render(state)`
- do we assume env is batched (and reset/step are vmapped under the hood) like Agent, or rather have the user (or rollout) do a `jax.vmap(env.step)()` call?
